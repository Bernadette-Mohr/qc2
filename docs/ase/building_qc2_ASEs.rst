.. _build_ASEs:

Building your own qc2-ASE calculator
====================================
