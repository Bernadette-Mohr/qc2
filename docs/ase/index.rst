ASE calculators
===============


.. toctree::
  :maxdepth: 1
