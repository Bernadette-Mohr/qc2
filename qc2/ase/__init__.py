"""qc2 ASE package."""
from .pyscf import PySCF
from .rose import ROSE, ROSETargetMolecule, ROSEFragment
from .dirac import DIRAC
