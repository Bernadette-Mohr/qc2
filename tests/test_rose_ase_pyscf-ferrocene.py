import os
import pytest

from .rose_test_functions import clean_up, extract_number, read_output

from ase import Atoms
from qc2.ase.rose import Rose
from qc2.ase.pyscf import PySCF


def run_ferrocene_rose_no_avas():
    """Ferrocene mol_frag/relativistic example calculation."""
    # define target molecule
    mol = Atoms('FeC10H10',
                positions=[
                    [ 1.66780866,  0.12282848, -0.27460150],
                    [ 2.24106869,  2.10024551, -0.33511480],
                    [ 3.31616817, -1.00607736, -0.77719710],
                    [ 1.78191728,  1.75217481,  0.97966357],
                    [ 1.16672810,  1.86346489, -1.25728635],
                    [ 2.24424364, -1.24911576, -1.70047730],
                    [ 2.85541376, -1.35028551,  0.53811026],
                    [ 0.42377202,  1.30026137,  0.87001390],
                    [ 0.04359901,  1.36901828, -0.51247557],
                    [ 1.12099915, -1.74357622, -0.95584975],
                    [ 1.49870213, -1.80608183,  0.42766268],
                    [ 3.23295333,  2.46614000, -0.58966882],
                    [ 4.30039797, -0.61794626, -1.02844385],
                    [ 2.36507558,  1.80777918,  1.89582556],
                    [ 1.20126378,  2.02039952, -2.33279589],
                    [ 2.27428727, -1.07886582, -2.77411019],
                    [ 3.42876748, -1.26708471,  1.45836098],
                    [-0.20304943,  0.95502659,  1.68878543],
                    [-0.92220706,  1.08645766, -0.92472956],
                    [ 0.15065240, -2.01282736, -1.36630431],
                    [ 0.86407561, -2.12909803,  1.24950292]
                    ],
                   calculator=PySCF(method='scf.RHF',
                                    basis='unc-sto-3g',
                                    multiplicity=1,
                                    relativistic='x2c',
                                    cart=True,
                                    scf_addons='frac_occ'
                                    )
                   )

    # define atomic/molecular fragments
    frag1 = Atoms('Fe',
                  positions=[[1.66780866, 0.12282848, -0.27460150]],
                  calculator=PySCF(method='scf.RHF',
                                   basis='unc-sto-3g',
                                   multiplicity=1,
                                   cart=True,
                                   relativistic=True,
                                   charge=2,
                                   scf_addons='frac_occ'
                                   )
                   )

    frag2 = Atoms('C5H5',
                positions=[
                    [ 2.24106869,  2.10024551, -0.33511480],
                    [ 1.78191728,  1.75217481,  0.97966357],
                    [ 1.16672810,  1.86346489, -1.25728635],
                    [ 0.42377202,  1.30026137,  0.87001390],
                    [ 0.04359901,  1.36901828, -0.51247557],
                    [ 2.36507558,  1.80777918,  1.89582556],
                    [ 1.20126378,  2.02039952, -2.33279589],
                    [-0.20304943,  0.95502659,  1.68878543],
                    [-0.92220706,  1.08645766, -0.92472956],
                    [ 3.23295333,  2.46614000, -0.58966882]
                    ],
                  calculator=PySCF(method='scf.RHF',
                                   basis='unc-sto-3g',
                                   multiplicity=1,
                                   cart=True,
                                   relativistic=True,
                                   charge=-1,
                                   scf_addons='frac_occ'
                                   )
                  )

    frag3 = Atoms('C5H5',
                  positions=[
                    [1.49870213, -1.80608183,  0.42766268],
                    [2.24424364, -1.24911576, -1.70047730],
                    [3.31616817, -1.00607736, -0.77719710],
                    [1.12099915, -1.74357622, -0.95584975],
                    [2.85541376, -1.35028551,  0.53811026],
                    [4.30039797, -0.61794626, -1.02844385],
                    [2.27428727, -1.07886582, -2.77411019],
                    [3.42876748, -1.26708471,  1.45836098],
                    [0.15065240, -2.01282736, -1.36630431],
                    [0.86407561, -2.12909803,  1.24950292]
                    ],
                  calculator=PySCF(method='scf.RHF',
                                   basis='unc-sto-3g',
                                   multiplicity=1,
                                   cart=True,
                                   relativistic=True,
                                   charge=-1,
                                   scf_addons='frac_occ'
                                   )
                  )

    mol.get_potential_energy()
    frag1.get_potential_energy()
    frag2.get_potential_energy()
    frag3.get_potential_energy()

    # Rose-ASE calculator
    rose_calc = Rose(rose_calc_type='mol_frag',
                     exponent=4,
                     rose_target=mol,
                     rose_frags=[frag1, frag2, frag3],
                     test = True,
                     relativistic = True
                     )
    
    # run the calculator
    rose_calc.calculate()


EXPECTED_OUTPUT_FILE = 'test_rose_ase_pyscf-ferrocene.stdout'
ACTUAL_OUTPUT_FILE = 'OUTPUT_ROSE'

CHARGE_REGEX = 'Partial charges of fragments.*\n{}'.format(('*\n.' * 6) + '*\n')
MO_ENERGIES_REGEX = 'Recanonicalized virtual energies of fragments.*\n{}'.format(('*\n.' * 85) + '*\n')
HF_ENERGIES_REGEX = 'HF energy .*\n{}'.format(('*\n.' * 2) + '*\n')


def run_calculation():
    """Calculation to run."""
    run_ferrocene_rose_no_avas()


@pytest.fixture(scope="session", autouse=True)
def clean_up_files():
    """Runs always at the end of each test."""
    yield
    clean_up()


def test_partial_charges():
    """Test case #1 - Partial charges of fragments."""
    # run the calculation
    run_calculation()

    # read the expected and actual outputs
    expected_output = read_output(os.path.join(os.path.dirname(__file__), 'data', EXPECTED_OUTPUT_FILE))
    actual_output = read_output(os.path.join(os.path.dirname(__file__), ACTUAL_OUTPUT_FILE))

    # check the actual output against the expected output
    expected_charges = extract_number(CHARGE_REGEX, expected_output)
    actual_charges = extract_number(CHARGE_REGEX, actual_output)
    assert actual_charges == pytest.approx(expected_charges, rel=1e-3)


def test_virtual_energies():
    """Test case #2 - Recanonicalized virtual energies of fragments."""
    # read the expected and actual outputs
    expected_output = read_output(os.path.join(os.path.dirname(__file__), 'data', EXPECTED_OUTPUT_FILE))
    actual_output = read_output(os.path.join(os.path.dirname(__file__), ACTUAL_OUTPUT_FILE))

    # check the actual output against the expected output
    expected_energies = extract_number(MO_ENERGIES_REGEX, expected_output)
    actual_energies = extract_number(MO_ENERGIES_REGEX, actual_output)
    assert actual_energies == pytest.approx(expected_energies, rel=1.0e-5)


def test_hf_energy():
    """Test case #3 - Check final HF energies, if test = True."""
    # read the expected and actual outputs
    expected_output = read_output(os.path.join(os.path.dirname(__file__), 'data', EXPECTED_OUTPUT_FILE))
    actual_output = read_output(os.path.join(os.path.dirname(__file__), ACTUAL_OUTPUT_FILE))

    # check the actual output against the expected output
    expected_energy = extract_number(HF_ENERGIES_REGEX, expected_output)
    actual_energy = extract_number(HF_ENERGIES_REGEX, actual_output)
    assert actual_energy == pytest.approx(expected_energy, rel=1.0e-5)
