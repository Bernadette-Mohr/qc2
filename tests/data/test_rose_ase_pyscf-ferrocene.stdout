 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @       Reduction of Orbital Space Extent        @
 @                     (ROSE)                     @
 @                                                @
 @ By Bruno Senjean, Souloke Sen, Michal Repisky, @
 @        Gerald Knizia and Lucas Visscher.       @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

 Select algebra (rcq) based on the use of restricted and spatial orbitals:
   rcq = 1, 2 or 4  : real, complex or quaternion algebra, respectively
   restricted       :  T
   spatial_orbitals :  T
   rcq              :  1

 ####################################
 #      Molecular  Coordinates      #
 ####################################

Atom    Z           X           Y           Z
Fe  26.00  3.15170160  0.23211219 -0.51892163
C    6.00  4.23500605  3.96888881 -0.63327519
C    6.00  6.26664962 -1.90121067 -1.46868966
C    6.00  3.36733564  3.31113051  1.85129584
C    6.00  2.20479657  3.52143828 -2.37592686
C    6.00  4.24100584 -2.36048668 -3.21343638
C    6.00  5.39594998 -2.55166980  1.01688102
C    6.00  0.80081306  2.45713788  1.64408800
C    6.00  0.08239019  2.58706961 -0.96843847
C    6.00  2.11838138 -3.29488153 -1.80629424
C    6.00  2.83213657 -3.41300002  0.80816534
H    1.00  6.10939637  4.66032918 -1.11431257
H    1.00  8.12657439 -1.16774919 -1.94347721
H    1.00  4.46934511  3.41620754  3.58259109
H    1.00  2.27005955  3.81800176 -4.40834534
H    1.00  4.29778007 -2.03876092 -5.24230850
H    1.00  6.47943148 -2.39444308  2.75590284
H    1.00 -0.38370781  1.80473870  3.19134195
H    1.00 -1.74271877  2.05310742 -1.74748561
H    1.00  0.28469178 -3.80369245 -2.58194095
H    1.00  1.63286625 -4.02341217  2.36121831

 ####################################
 #         Basis set size           #
 #               and                #
 #      Canonical MO Energies       #
 ####################################

 No. basis functions    B1:    237
 No. molecular orbitals B1:    218
 No. occupied  orbitals B1:     29
 No. molecular orbitals B2:     56


 ####################################
 #   Intrinsic Fragments Orbitals   #
 ####################################

  Version: Standard 2013

  --- Check Symmetrical Orthogonalization --- 
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.1510E-13


  --- Check Symmetrical Orthogonalization --- 
**SUCCESS** Symmetrical orthogonalization was successful, norm:  0.9590E-13

 ####################################
 #  Partial charges of fragments    #
 ####################################

       Fragment      Charge
   frag0           1.378767
   frag1          -0.689384
   frag2          -0.689384



 ####################################
 # Intrinsic Localized occupied MOs #
 ####################################


 ----- Percentage of localization before the localization procedure ----- 

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1     -1.1913 [    1    2    3]                                   
                [ 19.5 40.3 40.3]                                   
  2     -1.1704 [    2    3]                                        
                [ 50.0 50.0]                                        
  3     -0.9601 [    2    3]                                        
                [ 50.0 50.0]                                        
  4     -0.9601 [    2    3]                                        
                [ 50.0 50.0]                                        
  5     -0.9596 [    1    2    3]                                   
                [  1.2 49.4 49.4]                                   
  6     -0.9595 [    1    2    3]                                   
                [  1.2 49.4 49.4]                                   
  7     -0.7365 [    1    2    3]                                   
                [  1.7 49.2 49.2]                                   
  8     -0.7364 [    1    2    3]                                   
                [  1.7 49.2 49.2]                                   
  9     -0.7234 [    2    3]                                        
                [ 50.0 50.0]                                        
 10     -0.7234 [    2    3]                                        
                [ 50.0 50.0]                                        
 11     -0.7168 [    1    2    3]                                   
                [  6.9 46.6 46.6]                                   
 12     -0.6955 [    2    3]                                        
                [ 50.0 50.0]                                        
 13     -0.5636 [    1    2    3]                                   
                [ 12.2 43.9 43.9]                                   
 14     -0.5532 [    1    2    3]                                   
                [  1.8 49.1 49.1]                                   
 15     -0.5532 [    1    2    3]                                   
                [  1.8 49.1 49.1]                                   
 16     -0.5482 [    2    3]                                        
                [ 50.0 50.0]                                        
 17     -0.5482 [    2    3]                                        
                [ 50.0 50.0]                                        
 18     -0.5156 [    1    2    3]                                   
                [  1.8 49.1 49.1]                                   
 19     -0.5156 [    1    2    3]                                   
                [  1.8 49.1 49.1]                                   
 20     -0.5088 [    2    3]                                        
                [ 50.0 50.0]                                        
 21     -0.5088 [    2    3]                                        
                [ 50.0 50.0]                                        
 22     -0.4932 [    2    3]                                        
                [ 50.0 50.0]                                        
 23     -0.4594 [    1    2    3]                                   
                [ 91.6  4.2  4.2]                                   
 24     -0.3607 [    1    2    3]                                   
                [ 83.6  8.2  8.2]                                   
 25     -0.3605 [    1    2    3]                                   
                [ 83.5  8.2  8.2]                                   
 26     -0.3403 [    2    3]                                        
                [ 50.0 50.0]                                        
 27     -0.3403 [    2    3]                                        
                [ 50.0 50.0]                                        
 28     -0.3277 [    1    2    3]                                   
                [ 10.4 44.8 44.8]                                   
 29     -0.3276 [    1    2    3]                                   
                [ 10.4 44.8 44.8]                                   

 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        4.622565    7.49E-03
    2       25.468779    4.07E-03
    3       26.015220    1.15E-03
    4       26.045173    1.21E-04
    5       26.045477    3.57E-05
    6       26.045503    1.97E-07
    7       26.045503    1.61E-11
    8       26.045503    1.08E-13
    9       26.045503    1.24E-15
   10       26.045503    2.97E-17

 ******************************************
 Final converged values (it, func, fGrad):
   10           26.0455028527958206  2.9698E-17
 ******************************************


 ----- Percentage of localization after the localization procedure ----- 

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1]                                             
    [ 98.4]                                             
  2 [    1    2    3]                                   
    [ 87.1  6.5  6.5]                                   
  3 [    1    2    3]                                   
    [ 87.0  6.5  6.5]                                   
  4 [    2]                                             
    [100.0]                                             
  5 [    2]                                             
    [100.0]                                             
  6 [    2]                                             
    [100.0]                                             
  7 [    2]                                             
    [100.0]                                             
  8 [    2]                                             
    [100.0]                                             
  9 [    2]                                             
    [100.0]                                             
 10 [    2]                                             
    [100.0]                                             
 11 [    2]                                             
    [100.0]                                             
 12 [    2]                                             
    [100.0]                                             
 13 [    2]                                             
    [100.0]                                             
 14 [    1    2]                                        
    [ 15.8 83.4]                                        
 15 [    1    2]                                        
    [  6.7 93.1]                                        
 16 [    1    2]                                        
    [  6.7 93.1]                                        
 17 [    3]                                             
    [100.0]                                             
 18 [    3]                                             
    [100.0]                                             
 19 [    3]                                             
    [100.0]                                             
 20 [    3]                                             
    [100.0]                                             
 21 [    3]                                             
    [100.0]                                             
 22 [    3]                                             
    [100.0]                                             
 23 [    1    3]                                        
    [ 15.8 83.4]                                        
 24 [    3]                                             
    [100.0]                                             
 25 [    3]                                             
    [100.0]                                             
 26 [    3]                                             
    [100.0]                                             
 27 [    3]                                             
    [100.0]                                             
 28 [    1    3]                                        
    [  6.7 93.1]                                        
 29 [    1    3]                                        
    [  6.7 93.1]                                        

 ----- Percentage of localization after recanonization ----- 

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1     -0.4776 [    1]                                             
                [ 98.4]                                             
  2     -0.3711 [    1    2    3]                                   
                [ 87.1  6.5  6.5]                                   
  3     -0.3709 [    1    2    3]                                   
                [ 87.0  6.5  6.5]                                   
  4     -1.1796 [    1    2]                                        
                [  9.1 90.5]                                        
  5     -0.9598 [    2]                                             
                [ 99.4]                                             
  6     -0.9597 [    2]                                             
                [ 99.4]                                             
  7     -0.7263 [    2]                                             
                [100.0]                                             
  8     -0.7263 [    2]                                             
                [100.0]                                             
  9     -0.6973 [    2]                                             
                [100.0]                                             
 10     -0.5502 [    2]                                             
                [ 99.3]                                             
 11     -0.5502 [    2]                                             
                [ 99.3]                                             
 12     -0.5294 [    1    2]                                        
                [  6.7 92.9]                                        
 13     -0.5106 [    2]                                             
                [100.0]                                             
 14     -0.5106 [    2]                                             
                [100.0]                                             
 15     -0.3346 [    1    2]                                        
                [  5.4 94.4]                                        
 16     -0.3346 [    1    2]                                        
                [  5.4 94.4]                                        
 17     -1.1796 [    1    3]                                        
                [  9.1 90.5]                                        
 18     -0.9598 [    3]                                             
                [ 99.4]                                             
 19     -0.9597 [    3]                                             
                [ 99.4]                                             
 20     -0.7263 [    3]                                             
                [100.0]                                             
 21     -0.7263 [    3]                                             
                [100.0]                                             
 22     -0.6973 [    3]                                             
                [100.0]                                             
 23     -0.5502 [    3]                                             
                [ 99.3]                                             
 24     -0.5502 [    3]                                             
                [ 99.3]                                             
 25     -0.5294 [    1    3]                                        
                [  6.7 92.9]                                        
 26     -0.5106 [    3]                                             
                [100.0]                                             
 27     -0.5106 [    3]                                             
                [100.0]                                             
 28     -0.3346 [    1    3]                                        
                [  5.4 94.4]                                        
 29     -0.3346 [    1    3]                                        
                [  5.4 94.4]                                        


 ############################################
 # Intrinsic Localized valence virtual MOs  #
 ############################################


 ----- Percentage of localization before the localization procedure ----- 

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1      0.2477 [    1    2    3]                                   
                [ 66.8 16.6 16.6]                                   
  2      0.2482 [    2    3]                                        
                [ 50.0 49.9]                                        
  3      0.2483 [    2    3]                                        
                [ 49.9 49.9]                                        
  4      0.2484 [    2    3]                                        
                [ 49.9 49.9]                                        
  5      0.2484 [    2    3]                                        
                [ 49.4 50.4]                                        
  6      0.2749 [    2    3]                                        
                [ 50.0 49.9]                                        
  7      0.3032 [    2    3]                                        
                [ 54.9 45.1]                                        
  8      0.3032 [    2    3]                                        
                [ 11.9 88.1]                                        
  9      0.3196 [    2    3]                                        
                [ 79.3 20.7]                                        
 10      0.3196 [    2    3]                                        
                [ 67.1 32.9]                                        
 11      0.3210 [    2    3]                                        
                [ 48.2 51.8]                                        
 12      0.3213 [    2    3]                                        
                [ 58.7 41.3]                                        
 13      0.3521 [    2    3]                                        
                [ 62.0 38.0]                                        
 14      0.3584 [    2    3]                                        
                [ 57.0 43.0]                                        
 15      0.3588 [    2    3]                                        
                [ 60.5 39.5]                                        
 16      0.3614 [    2    3]                                        
                [ 29.1 70.9]                                        
 17      0.3614 [    2    3]                                        
                [ 25.4 74.6]                                        
 18      0.3852 [    2    3]                                        
                [ 46.3 53.7]                                        
 19      0.3853 [    2    3]                                        
                [ 51.2 48.8]                                        
 20      0.4591 [    2    3]                                        
                [ 48.3 50.8]                                        
 21      0.4591 [    2    3]                                        
                [ 50.1 49.9]                                        
 22      0.4808 [    2    3]                                        
                [ 49.9 50.1]                                        
 23      0.4808 [    1    2    3]                                   
                [  2.2 48.9 48.9]                                   
 24      0.5029 [    1    2    3]                                   
                [ 12.6 43.7 43.7]                                   
 25      0.5593 [    1    2    3]                                   
                [ 12.7 43.6 43.6]                                   
 26      0.6015 [    1    2    3]                                   
                [ 86.4  6.8  6.8]                                   
 27      0.6016 [    1    2    3]                                   
                [ 86.4  6.8  6.8]                                   

 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        5.481417    1.29E-02
    2       21.440123    1.12E-02
    3       24.241394    1.42E-03
    4       24.290614    7.54E-04
    5       24.299947    1.09E-04
    6       24.300135    1.66E-05
    7       24.300140    1.06E-07
    8       24.300140    1.47E-10
    9       24.300140    1.70E-12
   10       24.300140    1.36E-13
   11       24.300140    6.48E-15
   12       24.300140    4.64E-16

 ******************************************
 Final converged values (it, func, fGrad):
   12           24.3001395209418831  4.6409E-16
 ******************************************


 ----- Percentage of localization after the localization procedure ----- 

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1    2    3]                                   
    [ 68.3 15.8 15.8]                                   
  2 [    1    2    3]                                   
    [ 86.6  6.7  6.7]                                   
  3 [    1    2    3]                                   
    [ 86.6  6.7  6.7]                                   
  4 [    2]                                             
    [100.0]                                             
  5 [    2]                                             
    [100.0]                                             
  6 [    2]                                             
    [100.0]                                             
  7 [    2]                                             
    [ 99.9]                                             
  8 [    2]                                             
    [100.0]                                             
  9 [    2]                                             
    [100.0]                                             
 10 [    1    2]                                        
    [  6.5 93.4]                                        
 11 [    2]                                             
    [100.0]                                             
 12 [    2]                                             
    [ 99.9]                                             
 13 [    2]                                             
    [100.0]                                             
 14 [    1    2]                                        
    [  6.5 93.4]                                        
 15 [    2]                                             
    [ 99.2]                                             
 16 [    3]                                             
    [100.0]                                             
 17 [    3]                                             
    [100.0]                                             
 18 [    3]                                             
    [100.0]                                             
 19 [    3]                                             
    [100.0]                                             
 20 [    3]                                             
    [100.0]                                             
 21 [    3]                                             
    [100.0]                                             
 22 [    3]                                             
    [ 99.9]                                             
 23 [    3]                                             
    [100.0]                                             
 24 [    3]                                             
    [ 99.9]                                             
 25 [    3]                                             
    [ 99.2]                                             
 26 [    1    3]                                        
    [  6.5 93.4]                                        
 27 [    1    3]                                        
    [  6.5 93.4]                                        

 ####################################
 #      Additional virtuals         #
 ####################################

 No. molecular virtuals B3:    162
 No. selected  virtuals B3:     56
 No. reference virtuals B4:    121


 ####################################
 #   Intrinsic Fragments Orbitals   #
 ####################################

  Version: Standard 2013

  --- Check Symmetrical Orthogonalization --- 
**WARNING** Symmetrical orthogonalization has failed (thrsh: 1.D-12), norm:  0.1261E-10


  --- Check Symmetrical Orthogonalization --- 
**WARNING** Symmetrical orthogonalization has failed (thrsh: 1.D-12), norm:  0.9385E-10


 ##############################################
 # Intrinsic Localized additional virtual MOs #
 ##############################################


 ----- Localization procedure: 2x2 Rotations -----
 (Exponent = 4)

 Iter        Function    Gradient
    1        7.190552    2.90E-03
    2       44.669178    2.91E-03
    3       49.236740    7.13E-04
    4       49.484362    2.81E-04
    5       49.523244    6.29E-05
    6       49.524596    1.35E-05
    7       49.524653    7.36E-07
    8       49.524653    1.01E-09
    9       49.524653    5.36E-12
   10       49.524653    9.48E-14
   11       49.524653    2.35E-15
   12       49.524653    7.50E-17

 ******************************************
 Final converged values (it, func, fGrad):
   12           49.5246531761995001  7.5035E-17
 ******************************************


 ----- Percentage of localization after the localization procedure ----- 

 MO Localized on fragments (with %) (thrs = 0.01, first ten are shown)
  1 [    1    2    3]                                   
    [ 91.5  4.3  4.3]                                   
  2 [    1    2    3]                                   
    [ 94.7  2.6  2.6]                                   
  3 [    1    2    3]                                   
    [ 94.7  2.6  2.6]                                   
  4 [    1]                                             
    [ 98.8]                                             
  5 [    1]                                             
    [ 99.9]                                             
  6 [    2]                                             
    [100.0]                                             
  7 [    2]                                             
    [ 99.2]                                             
  8 [    2]                                             
    [100.0]                                             
  9 [    2]                                             
    [100.0]                                             
 10 [    2]                                             
    [100.0]                                             
 11 [    2]                                             
    [100.0]                                             
 12 [    2]                                             
    [100.0]                                             
 13 [    2]                                             
    [100.0]                                             
 14 [    1    2]                                        
    [  1.1 98.5]                                        
 15 [    2]                                             
    [100.0]                                             
 16 [    2]                                             
    [100.0]                                             
 17 [    2]                                             
    [100.0]                                             
 18 [    2]                                             
    [100.0]                                             
 19 [    1    2]                                        
    [  1.1 98.5]                                        
 20 [    1    2    3]                                   
    [ 13.3 82.6  4.1]                                   
 21 [    2]                                             
    [ 99.9]                                             
 22 [    2]                                             
    [100.0]                                             
 23 [    2]                                             
    [ 99.5]                                             
 24 [    2]                                             
    [ 99.5]                                             
 25 [    2]                                             
    [ 99.9]                                             
 26 [    2]                                             
    [ 99.9]                                             
 27 [    1    2    3]                                   
    [ 13.3 82.6  4.1]                                   
 28 [    1    2]                                        
    [  5.1 94.8]                                        
 29 [    1    2]                                        
    [  5.1 94.8]                                        
 30 [    1    2    3]                                   
    [ 21.8 69.6  8.5]                                   
 31 [    3]                                             
    [100.0]                                             
 32 [    3]                                             
    [100.0]                                             
 33 [    3]                                             
    [100.0]                                             
 34 [    3]                                             
    [100.0]                                             
 35 [    3]                                             
    [100.0]                                             
 36 [    3]                                             
    [ 99.5]                                             
 37 [    3]                                             
    [ 99.5]                                             
 38 [    3]                                             
    [100.0]                                             
 39 [    3]                                             
    [100.0]                                             
 40 [    3]                                             
    [ 99.2]                                             
 41 [    1    2    3]                                   
    [  2.6 48.7 48.7]                                   
 42 [    1    3]                                        
    [  1.1 98.5]                                        
 43 [    3]                                             
    [100.0]                                             
 44 [    3]                                             
    [100.0]                                             
 45 [    3]                                             
    [100.0]                                             
 46 [    1    2    3]                                   
    [ 13.3  4.1 82.6]                                   
 47 [    3]                                             
    [ 99.9]                                             
 48 [    3]                                             
    [100.0]                                             
 49 [    3]                                             
    [ 99.9]                                             
 50 [    3]                                             
    [ 99.9]                                             
 51 [    1    2    3]                                   
    [ 13.3  4.1 82.6]                                   
 52 [    1    3]                                        
    [  1.1 98.5]                                        
 53 [    1    3]                                        
    [  5.1 94.8]                                        
 54 [    1    3]                                        
    [  5.1 94.8]                                        
 55 [    3]                                             
    [100.0]                                             
 56 [    1    2    3]                                   
    [ 21.8  8.5 69.6]                                   

 ----- Recanonicalized virtual energies of fragments -----

    Fragment          MO              Energy
           1           1         0.378351611
           1           2         0.378352296
           1           3         0.458949306
           1           4         0.535888892
           1           5         0.734069297
           1           6         0.734073677
           1           7         0.770676430
           1           8         1.510927542
           2           1         0.250110145
           2           2         0.250226860
           2           3         0.265656611
           2           4         0.332701259
           2           5         0.332714416
           2           6         0.339462141
           2           7         0.339463700
           2           8         0.443919757
           2           9         0.443941122
           2          10         0.611069903
           2          11         0.611116994
           2          12         0.692453640
           2          13         0.701546099
           2          14         0.817068091
           2          15         0.817085366
           2          16         0.874043195
           2          17         0.874096400
           2          18         0.955490927
           2          19         1.025780809
           2          20         1.025788753
           2          21         1.130303467
           2          22         1.130712232
           2          23         1.199736248
           2          24         1.199798782
           2          25         1.203764531
           2          26         1.203776294
           2          27         1.303344154
           2          28         1.303403207
           2          29         1.305316726
           2          30         1.347802087
           2          31         1.448305070
           2          32         1.448551365
           2          33         1.482422713
           2          34         1.482468827
           2          35         1.727698367
           2          36         1.744216102
           2          37         1.744358390
           3           1         0.250110144
           3           2         0.250226861
           3           3         0.264873987
           3           4         0.332701253
           3           5         0.332714417
           3           6         0.339462141
           3           7         0.339463700
           3           8         0.443919747
           3           9         0.443940944
           3          10         0.611069899
           3          11         0.611116994
           3          12         0.648893981
           3          13         0.701546015
           3          14         0.787829524
           3          15         0.817080723
           3          16         0.817142022
           3          17         0.874043193
           3          18         0.874096404
           3          19         1.025780670
           3          20         1.025788308
           3          21         1.130303449
           3          22         1.130712218
           3          23         1.131747734
           3          24         1.199736246
           3          25         1.199798783
           3          26         1.203765975
           3          27         1.203777026
           3          28         1.303375128
           3          29         1.303403707
           3          30         1.306246446
           3          31         1.347802086
           3          32         1.448305069
           3          33         1.448551364
           3          34         1.482423097
           3          35         1.482468831
           3          36         1.744216096
           3          37         1.744358390
           3          38         1.817843957

 ####################################
 #         Starting TESTS           #
 ####################################

 The matrix is normalized.
 The matrix is normalized.
 The matrix is normalized.
 Deviation of IBO_RDM from HF_RDM  :   1.2960531874444305E-013
 Deviation from orthogonality      :   1.0433477936082955E-008
 Deviation from orthogonality      :   4.0618369383979275E-014
 Deviation in number of electrons  :  -2.1462653876369586E-010
 Deviation in number of electrons  :   0.0000000000000000     
 HF energy (moB1)   :   -1639.1395997879908     
 HF energy (IAO)    :   -1639.1395997900727     
 HF energy (IBO/MO) :   -1639.1395997879918     
 Deviation from the HF energy in the original basis and in the IAO basis:  -2.08183337E-09
 Deviation from the HF energy in the original basis and in the IBO basis:  -9.09494702E-13

 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
 @                                                @
 @       Reduction of Orbital Space Extent        @
 @                                                @
 @                      END                       @
 @                                                @
 @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
